xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,477 @@
+      checkArgument(!error.isOk(), "unexpected OK status");
+      if (closed) {
+        return;
+      }
+      closed = true;
+      cleanUp();
+      configWatcher.onError(error);
+      if (responseReceived || retryBackoffPolicy == null) {
+        // Reset the backoff sequence if had received a response, or backoff sequence
+        // has never been initialized.
+        retryBackoffPolicy = backoffPolicyProvider.get();
+      }
+      long delayNanos = 0;
+      if (!responseReceived) {
+        delayNanos =
+            retryBackoffPolicy.nextBackoffNanos() - stopwatch.elapsed(TimeUnit.NANOSECONDS);
+      }
+      logger.log(Level.FINE, "{0} stream closed, retry in {1} ns", new Object[]{this, delayNanos});
+      if (delayNanos <= 0) {
+        startDiscoveryRpc(channel);
+      } else {
+        rpcRetryTimer =
+            syncContext.schedule(
+                new RpcRetryTask(), delayNanos, TimeUnit.NANOSECONDS, timeService);
+      }
+    }
+
+    private void close(Exception error) {
+      if (closed) {
+        return;
+      }
+      closed = true;
+      cleanUp();
+      requestWriter.onError(error);
+    }
+
+    private void cleanUp() {
+      if (adsStream == this) {
+        adsStream = null;
+      }
+    }
+
+    private void sendLdsRequest(String resourceName, String nonce) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(ldsVersion)
+              .setNode(node)
+              .addResourceNames(resourceName)
+              .setTypeUrl(ADS_TYPE_URL_LDS)
+              .setResponseNonce(nonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+
+    private void sendRdsRequest(String resourceName, String nonce) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(rdsVersion)
+              .setNode(node)
+              .addResourceNames(resourceName)
+              .setTypeUrl(ADS_TYPE_URL_RDS)
+              .setResponseNonce(nonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+  }
+
+  /**
+   * Returns {@code true} iff {@code hostName} matches the domain name {@code pattern}.
+   *
+   * <p>Wildcard pattern rules:
+   * <ol>
+   * <li>A single asterisk (*) matches any domain.</li>
+   * <li>Asterisk (*) is only permitted in the left-most or the right-most part of the pattern,
+   *     but not both.</li>
+   * </ol>
+   */
+  private static boolean matchHostName(String hostName, String pattern) {
+    // Basic sanity checks
+    if (hostName == null || hostName.length() == 0 || hostName.startsWith(".")
+        || hostName.endsWith(".")) {
+      // Invalid domain name
+      return false;
+    }
+    if (pattern == null || pattern.length() == 0 || pattern.startsWith(".")
```

Comment:
I think `null` is never the case for protobuf message and the `hostName` field.

Reply:
True. This method is a static helper method, we could definitely make the spec stronger by weakening the precondition, as the check is not expensive.

Reply:
nit: Then you can treat the "precondition" as a `Precondition` instead of `return false`.

Reply:
If precondition fails, anything can happen. `return false` is just a benign choice.

Reply:
nit: Precondition failure as you said in another PR should "never ever reach". If it does reach, that means the code has a bug and should throw other than benignly return false.

Reply:
I am talking about specification vs implementation. Implementation can accept preconditions weaker than specified (aka, strengthen the specification), it still follows the spec and will not surprise users.